File: 19-Sqrt-Decomposition-MOs/02-MOs-Algorithm/java/MosAlgorithm.java
import java.util.*;
public class MosAlgorithm { static int[] cnt=new int[1000005]; static int curAns=0;
    static void add(int x){if(cnt[x]++==0)curAns++;}
    static void rem(int x){if(--cnt[x]==0)curAns--;}
    public static void main(String[] args){int n=7;int[] a={1,2,1,3,1,2,1};int block=(int)Math.sqrt(n);
    int[][] queries={{1,5,0},{0,3,1},{2,6,2}};
    Arrays.sort(queries,(x,y)->x[0]/block!=y[0]/block?x[0]/block-y[0]/block:x[1]-y[1]);
    int cl=0,cr=-1;int[] ans=new int[3];
    for(int[] q:queries){while(cr<q[1])add(a[++cr]);while(cl>q[0])add(a[--cl]);while(cr>q[1])rem(a[cr--]);while(cl<q[0])rem(a[cl++]);ans[q[2]]=curAns;}
    for(int i=0;i<3;i++)System.out.println("Query "+i+": "+ans[i]);}}

File: 19-Sqrt-Decomposition-MOs/02-MOs-Algorithm/cpp/mos_algorithm.cpp
/* MO's Algorithm - Distinct Elements in Range | Time: O((N+Q)*sqrt(N)) */
#include <bits/stdc++.h>
using namespace std;
int cnt[1000005],curAns=0,block;
struct Query{int l,r,idx;};
void add(int x){if(cnt[x]++==0)curAns++;}
void rem(int x){if(--cnt[x]==0)curAns--;}
int main(){int n=7;int a[]={1,2,1,3,1,2,1};block=(int)sqrt(n);
vector<Query>queries={{1,5,0},{0,3,1},{2,6,2}};
sort(queries.begin(),queries.end(),[](auto&a,auto&b){return a.l/block!=b.l/block?a.l/block<b.l/block:a.r<b.r;});
int cl=0,cr=-1;vector<int>ans(3);
for(auto&q:queries){while(cr<q.r)add(a[++cr]);while(cl>q.l)add(a[--cl]);while(cr>q.r)rem(a[cr--]);while(cl<q.l)rem(a[cl++]);ans[q.idx]=curAns;}
for(int i=0;i<3;i++)cout<<"Query "<<i<<": "<<ans[i]<<" distinct"<<endl;}
